homeassistant/components/dlna_dms/dms.py

```diff
@@ -0,0 +1,726 @@
+        while (suggested_source_id := f"{source_id_base}_{tries}") in self.sources:
+            tries += 1
+
+        return suggested_source_id
+
+
+@callback
+def get_domain_data(hass: HomeAssistant) -> DlnaDmsData:
+    """Obtain this integration's domain data, creating it if needed."""
+    if DOMAIN in hass.data:
+        return cast(DlnaDmsData, hass.data[DOMAIN])
+
+    data = DlnaDmsData(hass)
+    hass.data[DOMAIN] = data
+    return data
+
+
+@dataclass
+class DidlPlayMedia(PlayMedia):
+    """Playable media with DIDL metadata."""
+
+    didl_metadata: didl_lite.DidlObject
+
+
+class DlnaDmsDeviceError(BrowseError, Unresolvable):
+    """Base for errors raised by DmsDeviceSource.
+
+    Caught by both media_player (BrowseError) and media_source (Unresolvable),
+    so DmsDeviceSource methods can be used for both browse and resolve
+    functionality.
+    """
+
+
+class DeviceConnectionError(DlnaDmsDeviceError):
+    """Error occurred with the connection to the server."""
+
+
+class ActionError(DlnaDmsDeviceError):
+    """Error when calling a UPnP Action on the device."""
+
+
+def catch_request_errors(
+    func: Callable[[_DlnaDmsDeviceMethod, str], Coroutine[Any, Any, _RetType]]
+) -> Callable[[_DlnaDmsDeviceMethod, str], Coroutine[Any, Any, _RetType]]:
+    """Catch UpnpError errors."""
+
+    @functools.wraps(func)
+    async def wrapper(self: _DlnaDmsDeviceMethod, req_param: str) -> _RetType:
+        """Catch UpnpError errors and check availability before and after request."""
+        if not self.available:
+            LOGGER.warning("Device disappeared when trying to call %s", func.__name__)
+            raise DeviceConnectionError("DMS is not connected")
+
+        try:
+            return await func(self, req_param)
+        except UpnpActionError as err:
+            LOGGER.debug("Server failure", exc_info=err)
+            if err.error_code == ContentDirectoryErrorCode.NO_SUCH_OBJECT:
+                LOGGER.debug("No such object: %s", req_param)
+                raise ActionError(f"No such object: {req_param}") from err
+            if err.error_code == ContentDirectoryErrorCode.INVALID_SEARCH_CRITERIA:
+                LOGGER.debug("Invalid query: %s", req_param)
+                raise ActionError(f"Invalid query: {req_param}") from err
+            raise DeviceConnectionError(f"Server failure: {err!r}") from err
+        except UpnpConnectionError as err:
+            LOGGER.debug("Server disconnected", exc_info=err)
+            await self.device_disconnect()
+            raise DeviceConnectionError(f"Server disconnected: {err!r}") from err
+        except UpnpError as err:
+            LOGGER.debug("Server communication failure", exc_info=err)
+            raise DeviceConnectionError(
+                f"Server communication failure: {err!r}"
+            ) from err
+
+    return wrapper
+
+
+class DmsDeviceSource:
+    """DMS Device wrapper, providing media files as a media_source."""
+
+    hass: HomeAssistant
```

Comment:
We only need to type attributes here if it differs from the type annotated in the init method for the attributes.